lib/ecto/multi.ex

```diff
@@ -303,10 +304,11 @@ defmodule Ecto.Multi do
   Adds a function to run as part of the multi.
 
   The function should return either `{:ok, value}` or `{:error, value}`,
-  and receives changes so far as an argument.
+  and receives the repo as the first argument, and the changes so far
+  as the second argument.
   """
   @spec run(t, name, (t -> {:ok | :error, any})) :: t
-  def run(multi, name, run) when is_function(run, 1) do
+  def run(multi, name, run) when is_function(run, 2) do
```

Comment:
Can't this be backwards compatible with 2.x by accepting either a 1-arity or 2-arity function?

Reply:
Anonymous functions and MFA functions should have the same signature.

While anonymous functions are _guaranteed_ to not have additional arguments, and thus we can test for the arity, doing the same for MFA can be ambiguous.

Example:

```
defmodule SomeModule do
  def some_function(changes, arg1 \\ nil) do
    {:ok, true}
  end 
end

multi1 = Multi.new |> Multi.run(:fun, SomeModule, :some_function, [arg1])
multi2 = Multi.new |> Multi.run(:fun, SomeModule, :some_function, [])
```

In this contrived example, SomeModule has a function, `some_function`, which has both arity 1 and 2.  Do we:

  * Check for arity 2 first and pass changes, repo? In this case we'd be wrong for multi2.
  * Check for arity 1 first and pass just changes? Can't do that.
  * Check for arity 2 + args length first? This is getting complicated.

While we could probably come up with something clever, I'm not sure if the cleverness would cause more confusing problems for some people than just documenting the upgrade path from Ecto 2.x to 3.x.

More than happy to hear any suggestions.

Reply:
With MFA sure, but the feature of passing in `repo` now could be done via anonymous functions.  The MFA form can be added/changed in 3.0.  It is not much of a cost to wrap an MFA in an anonymous function anyway and it adds the feature in a backwards compatible way too.

Reply:
Plus you could even use `:erlang.function_exported`(sp?) to test the mfa and args to see if it is exported (and if both are then throw an error saying to disambiguate via an anonymous function or so).

Or just add a new :run method with a new argument or name or so...

Reply:
I'll wait for more direction from a maintainer as to how/if they'd like to proceed and how they want to handle backward compatibility.

Valid argument & suggestions though. Thanks!

Reply:
We can emit  a warning for anonymous functions like this:

```elixir
def run(multi, name, run) when is_function(run, 1) do
  IO.warn "Giving an anonymous function with 1 argument to Ecto.Multi.run/3 is deprecated. " <>
            "Please pass an anonymous function that receives the repository and the changes so far"
  run(multi, name, fn _repo, changes -> run.(changes) end)
end
```

Given it is only 5 LOC and it will lead to a better upgrading experience, we should do it, yeah.